### PR TITLE
Add Trello extension

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
## Description

This pull request adds a draft of a trello extension, with a few questions, mostly as I am learning React…

## Type of change

- New extension

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

### CODE REVIEW

File: extensions/trello/src/TrelloListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  export const TodoListItem = ({
+    result,
+  }: TodoListItemProps): JSX.Element => {
+    const todo = result
+  
+    return (
+      <List.Item
+        id={todo.id}
+        title={todo.name}
+        // accessoryTitle={todo.due?.toLocaleDateString()}
```
Comment (ChrisChinchilla) [inline extensions/trello/src/TrelloListItem.tsx:22]: How do you use a function like this when the value could sometimes be empty?

File: extensions/trello/src/TrelloListItem.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        subtitle={todo.desc}
+        icon={Icon.Checkmark}
+        // TODO: How to do the following?
+        // keywords={todo.labels.name}
```
Comment (ChrisChinchilla) [inline extensions/trello/src/TrelloListItem.tsx:26]: I noticed some extensions that seem to popular this value automatically, but with the Trello API, labels are nested objects inside the main card object, not sure how to return all name values of the label objects…

File: extensions/trello/src/utils/fetchTodos.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { getPreferenceValues, showToast, ToastStyle } from '@raycast/api'
+import fetch from 'node-fetch'
+import { TrelloFetchResponse } from '../trelloResponse.model'
+const TRELLO_APP_KEY = "13f8c59607ba6d82531d3db5f46999c1";
```
Comment (ChrisChinchilla) [inline extensions/trello/src/utils/fetchTodos.tsx:4]: This is OK to be public

**Comment by ChrisChinchilla** [comment]
And a general question… How can I create an action that sends something back to the Trello API?

File: extensions/trello/src/TrelloListItem.tsx (hunk 2 above)
Comment (Ali-Parandeh) [reply extensions/trello/src/TrelloListItem.tsx:26]:
```typescript
todo?.labels?.map(label => label.toUpperCase())
```

File: extensions/trello/src/utils/fetchTodos.tsx (hunk 3 above)
Comment (Ali-Parandeh) [reply extensions/trello/src/utils/fetchTodos.tsx:4]: Should use preferences

in package.json
```json
"preferences": [
    {
      "name": "token",
      "type": "password",
      "required": true,
      "title": "API Token",
      "description": "Your Trello API token taken from xxxx",
      "link": "https://github.com/xxx/xxxx",
      "placeholder": "Enter your Trello API token."
    }
]
...

then in utils.ts

```typescript
import { getPreferenceValues } from "@raycast/api";

interface Preferences {
  token: string;
  teamId: string;
  spaceId: string;
}

const preferences: Preferences = getPreferenceValues();

export default preferences;

```

**Comment by Ali-Parandeh** [comment]
@ChrisChinchilla Use a useEffect

```typescript
function useMycustomPostRequest(data){
  const [response, setResponse] = useState(null);


  useEffect(() => {
    async function postData(postData){
      try {
        const response = await axios.post(url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${preferences.token}`
          },
          data: postData
        });
        setResponse(response);
      } catch (error) {
        console.log(error);
      }
    }

    postData(data).then((r) => r);
  }, [data])

  return response;

}

export { useMycustomPostRequest };
```

File: extensions/trello/src/TrelloListItem.tsx (hunk 1 above)
Comment (Ali-Parandeh) [reply extensions/trello/src/TrelloListItem.tsx:22]:
```typescript
id = {todo?.id ?? "Fallblack Text"}
```

**Comment by ChrisChinchilla** [comment]
Thanks so much @Ali-Parandeh will apply!

File: extensions/trello/src/TrelloListItem.tsx (hunk 1 above)
Comment (ChrisChinchilla) [reply extensions/trello/src/TrelloListItem.tsx:22]: @Ali-Parandeh Hmm, but does that apply here? To be more specific I mean `{todo.due.toLocaleDateString()}`, the `.toLocaleDateString()` sometimes results in error if there is no value, i.e. "TypeError: todo.due?.toLocaleDateString is not a function"

Comment (ChrisChinchilla) [reply extensions/trello/src/TrelloListItem.tsx:22]: I cannot get this to work, it's optional right now anyway

Comment (thomaspaulmann) [reply extensions/trello/src/TrelloListItem.tsx]: Is the `due` a Date object?

Comment (ChrisChinchilla) [reply extensions/trello/src/TrelloListItem.tsx]: @thomaspaulmann well I made it one in the 'model', but the Trello API returns a string, maybe I was doing something wrong there

Comment (thomaspaulmann) [reply extensions/trello/src/TrelloListItem.tsx]: That could be the case. It depends on what it returns. Best to use string and convert it if you need to.

**Comment by sxn** [comment]
hey @ChrisChinchilla 👋 just wanted to ask what the status on this is - is this ready to be reviewed or do you still need some time to work on it?

**Comment by ChrisChinchilla** [comment]
Hey @sxn I am still having some issues with test failing (they pass locally), but I think they relate to an issue I am having with Typescript models as the Trello API returns different API responses with the two endpoints I am using, making it hard for me to make it er, "happy", although things all work… I posted a message in the Slack channel about it, but basically one endpoint returns a list of cards, the other returns the list inside a "cards" array, and I am not sure how to solve it…

**Comment by ChrisChinchilla** [comment]
@sxn I'm stumped on a better way to fix the issues with Typescript and the different API responses, but as part of recent changes, I fixed code that means if you changed a search term, things would crash and the only way I could find to resolve things, was return the object of the response and add an ignore to the typescript lint… It's not satisfactory, but everything works 🤷‍♂️

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @ChrisChinchilla 👋 

Thanks for your contribution 💯 🔥 

I have some feedback for you to look into 💪 

- Your extension have [Empty state flicker](https://developers.raycast.com/basics/prepare-an-extension-for-store#empty-states) - Could you look into this?


- A better README would increase the UX a lot, I was confused about the naming of the preferences and what to put where since in Trello its just called `Key` (and API Token in preferences)

N1FolUPy0fABwikC


- When opening the extension I can see one item, when starting to search it disappears - but it depends on what I'm search for – also note a lot more on the list.

14e3hu7ZOzXMFb6K

qfC9CRvmn9tQhngp

dGXIHOD1yDDR9i5S

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by ChrisChinchilla** [comment]
@pernielsentikaer The read me I knew and kind of forgot about. I am a React and Typescript newbie, but I’ll attempt to understand what these things mean. A lot of the code came from other extensions, so they could also have the same issues…

**Comment by ChrisChinchilla** [comment]
@pernielsentikaer Regarding the empty state issue, as far as I can tell (and this code mostly came from the npm package search), I have implemented what you suggest?

From here down…

https://github.com/raycast/extensions/blob/c78aa513cd296900ca17967864eb3fce4e780633/extensions/trello/src/index.tsx#L36

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @ChrisChinchilla 👋 

Thanks for your time doing this, check the suggestions which should fix the [empty state flicker](https://developers.raycast.com/basics/prepare-an-extension-for-store#empty-states)

Can't remember if you looked into why the initial list only shows one, but if searching, It's possible to find more?

Looking forward to testing it out again 💯 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/trello/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { List, showToast, ToastStyle } from "@raycast/api";
+import { useEffect, useState } from "react";
+import { returnTodos } from "./utils/fetchTodos";
+import { TrelloFetchResponse } from "./trelloResponse.model";
+import { TodoListItem } from "./TrelloListItem";
+
+export default function PackageList() {
+  const [results, setTodos] = useState<TrelloFetchResponse>([]);
+  const [loading, setLoading] = useState<boolean>(false);
```
Comment (pernielsentikaer) [inline extensions/trello/src/index.tsx:9]:
```suggestion
  const [loading, setLoading] = useState<boolean>(true);
```
Set the initial state to true to completely avoid flicker, forgot that one 😄

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @ChrisChinchilla 👋 

I forgot the last correction, could you add that also?

Is it intended, that first view shows only my cases, and I can search in all cases in the boards I have? (as the screenshots earlier)

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/trello/package.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      "mode": "view"
+    }
+  ],
+  "preferences": [
+    {
+      "name": "username",
+      "type": "textfield",
+      "required": true,
+      "title": "Trello username",
+      "description": "Trello username",
```
Comment (pernielsentikaer) [inline extensions/trello/package.json:24]:
```suggestion
      "title": "Trello Username",
      "description": "Trello Username",
```
To match the style in other preferences

**Comment by ChrisChinchilla** [comment]
> Is it intended, that first view shows only my cases, and I can search in all cases in the boards I have? (as the screenshots earlier)


I don't fully understand the question. But If I understand correctly, then yes, at the moment the extension returns all cards assigned to you, and offers the ability to search the text of those. There will be further improvements soon though :)

**Comment by pernielsentikaer** [comment]
@ChrisChinchilla I'm just a bit wondering why I only see one when loading the extension, and when I search there is alot more:

image

image

Maybe it's as expected, but as a user I would be pretty confused 😄

**Comment by ChrisChinchilla** [comment]
@pernielsentikaer erm, that’s not what I see at all, as you can see from my early screen grab, it works filtering across a wider list. It could maybe be because that’s the only card assigned to you, and search is searching all cards across all the boards you’re on. I am away from my laptop until Monday afternoon so can’t check until then.

**Comment by pernielsentikaer** [comment]
It could be awesome if the initial load showed all instead it's confusing that I only see the one assigned directly to me, but I'm able to search for a lot more if I know what to search for

Could they be combined in one list with sections instead?

**Comment by ChrisChinchilla** [comment]
@pernielsentikaer OK, been looking into this and it seems unfortunately the way the Trello API is designed, others asking similar questions here - https://stackoverflow.com/questions/26552278/trello-api-getting-boards-lists-cards-information

In that there's an API for returning all cards assigned to you, and then an API for searching everything, but in the "search everything" API you can't filter by member… It's odd I agree, but that's the way it is 🤷 

There might be a way to get it to work as Intend by using react state or something, but that would involve a significant rewrite and digging into topics I don't fully understand right now. So how about in the spirits of getting this merged, I cut the search feature, and stick with the simple "return all your cards" for now, and add that back when I've had time to investigate?

**Comment by pernielsentikaer** [comment]
How about splitting the things apart?

1. Filter your cards, as today but but search, only filtering
2. Search for cards, starts with an empty array 

That way it should be solved 😄

**Review by pernielsentikaer — COMMENTED** [review]
Hi @ChrisChinchilla 👋 

Is there anything I can help you with in order to get it closed 😄

**Comment by ChrisChinchilla** [comment]
@pernielsentikaer Sorry it's been a very busy few weeks, not much time for open source work. I don't fully understand what you mean sorry…

1. Filter how? With the API? With Raycast/React? Filter on what?
2. If you mean just a "relevant Trello-wide" search, I still don't really feel like that actually makes much sense from the perspective of a personal tool like RayCast, but until i find a better way to get around that API limitation, I guess it will do…

**Comment by ChrisChinchilla** [comment]
OK, functionality split is done. I think i know what you mean regarding filtering, but I wonder what to filter on, there's potentially a lot of options…

File: extensions/trello/package.json [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  "name": "trello",
+  "title": "Trello",
+  "description": "Raycast extension for showing, searching, and interacting with cards from Trello",
+  "icon": "command-icon.png",
+  "author": "ChrisChinchilla",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "index",
+      "title": "Fetch Todos",
```
Comment (pernielsentikaer) [inline extensions/trello/package.json:12]:
```suggestion
      "title": "Fetch Assigned Todos",
```

File: extensions/trello/src/utils/fetchTodos.tsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      return json.cards as TrelloFetchResponse;
+    } else {
+      try {
+        const response = await fetch(
+          `https://api.trello.com/1/members/${username}/cards?filter=visible&key=${apitoken}&token=${token}`
+        );
+        const json = await response.json();
+        return json as TrelloFetchResponse;
+      } catch (error) {
+        showToast(ToastStyle.Failure, "Could not fetch ToDos");
```
Comment (pernielsentikaer) [inline extensions/trello/src/utils/fetchTodos.tsx:32]:
```suggestion
        showToast(ToastStyle.Failure, "An error occured", "Could not fetch todos, check your credentials");
```

File: extensions/trello/src/utils/fetchTodos.tsx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        return Promise.resolve([]);
+      }
+    }
+  } catch (error) {
+    showToast(ToastStyle.Failure, "Could not fetch ToDos");
```
Comment (pernielsentikaer) [inline extensions/trello/src/utils/fetchTodos.tsx:37]:
```suggestion
    showToast(ToastStyle.Failure, "An error occured", "Could not fetch todos, check your credentials");
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Awesome @ChrisChinchilla 🔥 

It was exactly what I was thinking about, sorry for not being clear enough!

A few suggestions, and then we're ready to ship it 🚢 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by ChrisChinchilla** [comment]
Thanks @pernielsentikaer rerequested!

**Comment by pernielsentikaer** [comment]
Awesome @ChrisChinchilla 🔥 

Approved! 😄

**Review by pernielsentikaer — APPROVED** [review]

**Comment by ChrisChinchilla** [comment]
🙀 🙀 🙀 🙀 🚀 🚀 🚀 🚀

File: extensions/trello/src/utils/fetchTodos.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import fetch from "node-fetch";
+import { TrelloFetchResponse } from "../trelloResponse.model";
+
+export interface Preferences {
+  token: string;
+  apitoken: string;
+  username: string;
+}
+
+const { token }: Preferences = getPreferenceValues();
```
Comment (fe9lix) [inline extensions/trello/src/utils/fetchTodos.tsx:11]: Hey @ChrisChinchilla – some optimization for a future version. You can destructure this in one go: const { token, apitoken, username } = ...

**Pull request merged**